sympy/matrices/expressions/vandermonde.py

```diff
@@ -0,0 +1,131 @@
+        least squares fitting, where each row corresponds to a polynomial point
+        and each column corresponds to a power of that point.
+
+        A square Vandermonde matrix for the points `x_1` to `x_n` is
+         .. math ::
+             \begin{bmatrix}
+             1 & x_1 & x_1^2 & \hdots % x_1^{n-1} \\
+             1 & x_2 & x_2^2 & \hdots % x_2^{n-1} \\
+            \vdots & \vdots & \vdots & \ddots & \vdots \\
+             1 & x_n & x_n^2 & \hdots % x_n^{n-1} \\
+             \end{bmatrix}
+
+        Parameters
+        ==========
+
+        points : iterable
+            The points to generate the VandermondeMatrix for.
+
+        n : integer, iterable or None
+            Powers to use. For an integer, the powers range from 0 to n-1.
+            For an iterable, use those powers. If ``None`` (default), the
+            powers range from 0 to ``len(points)``-1.
+
+        Examples
+        ========
+
+        >>> from sympy.matrices.expressions.vandermonde import VandermondeMatrix
+        >>> v = VandermondeMatrix([0, 1, -1])
+        >>> v.as_explicit()
+        Matrix([
+        [1,  0, 0],
+        [1,  1, 1],
+        [1, -1, 1]])
+
+        >>> from sympy import symbols
+        >>> x = symbols('x(1:4)')
+        >>> VandermondeMatrix(x, range(4)).as_explicit()
+        Matrix([
+        [1, x1, x1**2, x1**3],
+        [1, x2, x2**2, x2**3],
+        [1, x3, x3**2, x3**3]])
+
+        Sometimes, a Vandermonde matrix is defined without the power of zero
+        column:
+
+        >>> VandermondeMatrix(x, range(1,4)).as_explicit()
+        Matrix([
+        [x1, x1**2, x1**3],
+        [x2, x2**2, x2**3],
+        [x3, x3**2, x3**3]])
+
+        As it is possible to specify an iterable for the powers, generalized
+        versions can be rapildy created:
+
+        >>> VandermondeMatrix([-1, 2, 3], [1, 3, 6]).as_explicit()
+        Matrix([
+        [-1, -1,   1],
+        [ 2,  8,  64],
+        [ 3, 27, 729]])
+
+        It is also possible to use a symbolic column count:
+        >>> n = symbols('n')
+        >>> VandermondeMatrix([-1, 2, 3], n)
+        VandermondeMatrix((-1, 2, 3), n)
+
+        or symbolic powers
+
+        >>> VandermondeMatrix(x, [0, 1, n]).as_explicit()
+        Matrix([
+        [1, x1, x1**n],
+        [1, x2, x2**n],
+        [1, x3, x3**n]])
+
+        References
+        ==========
+
+        .. [1] https://en.wikipedia.org/wiki/Vandermonde_matrix
+
+    """
+    def __new__(cls, points, n=None):
+        if not iterable(points):
+            raise ValueError("First argument must be an iterable")
+        points = sympify(tuple(points))
+        n = sympify(n)
+        if n:
+            if iterable(n):
+                n = tuple(n)
+                cls.colpows = n
+            else:
+                cls._check_dim(n)
+                cls.colpows = Range(n)
+        else:
+            cls.colpows = Range(len(points))
+
+        if n is None:
+            obj = super().__new__(cls, points)
```

Comment:
It's better to normalize the argument 'n' to be fixed tuple of `(0, 1, 2, ...)` before stored because it makes easier to parse this later.

Reply:
The suggestion I would have is always try to normalize the arguments to have the same structure, such that both of them are explicit `Tuple, Tuple` at the best
And it works only except for the case where you have introduced symbolic `n` because Range can't compute the cardinality when `n` is symbolic

```
--- a/sympy/matrices/expressions/vandermonde.py
+++ b/sympy/matrices/expressions/vandermonde.py
@@ -1,5 +1,5 @@
 from sympy.core.compatibility import iterable
-from sympy.core.sympify import sympify
+from sympy.core.sympify import _sympify
 from sympy.matrices.expressions import MatrixExpr, Determinant
 from sympy import S, Pow, Range

@@ -90,33 +90,31 @@ class VandermondeMatrix(MatrixExpr):
     def __new__(cls, points, n=None):
         if not iterable(points):
             raise ValueError("First argument must be an iterable")
-        points = sympify(tuple(points))
-        n = sympify(n)
-        if n:
-            if iterable(n):
-                n = tuple(n)
-                cls.colpows = n
-            else:
-                cls._check_dim(n)
-                cls.colpows = Range(n)
-        else:
-            cls.colpows = Range(len(points))

-        if n is None:
-            obj = super().__new__(cls, points)
+        points = tuple(points)
+        points = _sympify(points)
+
+        if iterable(n):
+            n = _sympify(tuple(n))
+        elif n is None:
+            n = Range(len(points))
         else:
-            obj = super().__new__(cls, points, n)
-        return obj
+            n = _sympify(n)
+            if n.is_integer is False:
+                raise ValueError
+            if n.is_positive is False:
+                raise ValueError
+            n = Range(n)
+
+        return super().__new__(cls, points, n)
+
+    @property
+    def colpows(self):
+        return self.args[1]

     @property
     def shape(self):
-        rows = len(self.args[0])
-        if len(self.args) == 1 or self.args[1] is None:
-            return (rows, rows)
-        elif iterable(self.args[1]):
-            return (rows, len(self.args[1]))
-        else:
-            return (rows, self.args[1])
+        return len(self.args[0]), len(self.args[1])

     def _entry(self, i, j, **kwargs):
         return Pow(self.args[0][i], self.colpows[j])
(END)
```

Reply:
I see your point for sure! It may be that the cardinality actually can be computed now with symbolic `Range`.